<a href="https://colab.research.google.com/github/Scor999/Data_collection_and_markup/blob/main/dz9.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Задание 1.        
Выберите датасет, который имеет отношение к вашей области интересов или исследований. Датасет должен содержать неструктурированные данные, требующие разметки для решения конкретной задачи, например, анализа настроений или распознавания именованных сущностей.    
используем датасет https://www.kaggle.com/datasets/lakshmi25npathi/imdb-dataset-of-50k-movie-reviews/data

!!! Все файлы хранятся в каталоге dz9 репозиторя!!!!

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [ ]:
imdb_dataset = pd.read_csv('IMDB Dataset.csv')
imdb_dataset.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

Задание 2.    
Выполните разметку на основе правил (rule-based labeling) на подмножестве выбранного датасета. Разработайте и реализуйте набор правил или условий, которые позволят автоматически присваивать метки данным на основе определенных шаблонов или критериев.

In [ ]:
def rule_based_labeling(review):
    positive_keywords = ["wonderful", "amazing", "excellent", "great", "fantastic"]
    negative_keywords = ["bad", "terrible", "awful", "horrible", "disappointing"]

    # Приведем обзор к нижнему регистру для удобства сравнения
    review = review.lower()

    # Проверяем наличие ключевых слов
    if any(keyword in review for keyword in positive_keywords):
        return "positive"
    elif any(keyword in review for keyword in negative_keywords):
        return "negative"
    else:
        return "neutral"  # В случае отсутствия ключевых слов "neutral"

# Применим функцию к столбцу с обзорами и создадим новый столбец с метками, полученными на основе правил
imdb_dataset['rule_based_sentiment'] = imdb_dataset['review'].apply(rule_based_labeling)

In [ ]:
# Выведем первые несколько строк с новым столбцом
imdb_dataset[['review', 'sentiment', 'rule_based_sentiment']].head()

review sentiment  \
0  One of the other reviewers has mentioned that ...  positive   
1  A wonderful little production. <br /><br />The...  positive   
2  I thought this was a wonderful way to spend ti...  positive   
3  Basically there's a family where a little boy ...  negative   
4  Petter Mattei's "Love in the Time of Money" is...  positive   

  rule_based_sentiment  
0              neutral  
1             positive  
2             positive  
3              neutral  
4              neutral

In [ ]:
# Разделим датасет для ручной разметки
train_data, manual_labeling_data = train_test_split(imdb_dataset, test_size=0.001, random_state=42)

# Выведем информацию о размере каждой части
print("Размер обучающего набора данных:", len(train_data))
print("Размер данных для ручной разметки:", len(manual_labeling_data))


Размер обучающего набора данных: 49950
Размер данных для ручной разметки: 50


In [ ]:
# Сохраняем данные для разметки  в файл CSV
manual_labeling_data.to_csv('manual_labeling_data.csv', index=False)
train_data.to_csv('auto_manual.csv', index=False)


Задача 3.   
Выполните разметку вручную отдельного подмножества выбранного датасета с помощью выбранного вами инструмента разметки.

Далее установим библиотеку label-studio с помощью команды в терминале pip install label-studio и терминале же запустим команду label-studio

> Скриншоты находятся в каталоге dz9




In [ ]:
# Ручная разметка датасета
manual = pd.read_csv('manual_data.csv')

In [ ]:
manual

annotation_id  annotator                   created_at   id  lead_time  \
0               7          1  2024-04-06T07:17:11.963918Z  101      3.023   
1               3          1  2024-04-06T07:15:03.047730Z  102      3.295   
2               5          1  2024-04-06T07:16:52.574249Z  103      3.135   
3               6          1  2024-04-06T07:17:02.618572Z  104      4.258   
4               8          1  2024-04-06T07:17:19.914771Z  105      3.344   
5               9          1  2024-04-06T07:17:33.524727Z  106      2.054   
6               4          1  2024-04-06T07:16:14.321122Z  107      4.408   
7              10          1  2024-04-06T07:17:51.658532Z  108      5.254   
8              11          1  2024-04-06T07:18:00.669529Z  109      2.891   
9              12          1  2024-04-06T07:18:15.617215Z  110      2.343   
10             13          1  2024-04-06T07:18:26.905062Z  111      2.552   
11             14          1  2024-04-06T07:18:37.862661Z  112      2.730   
12             15          1  2024-04-06T07:18:47.374472Z  113      2.258   
13             16          1  2024-04-06T07:19:01.749669Z  114      2.228   
14             17          1  2024-04-06T07:19:09.190582Z  115      2.226   
15             18          1  2024-04-06T07:19:18.058440Z  116      1.902   
16             19          1  2024-04-06T07:19:27.389950Z  117      2.943   
17             20          1  2024-04-06T07:19:40.845416Z  118      3.046   
18             21          1  2024-04-06T07:19:59.331342Z  119      3.583   
19             22          1  2024-04-06T07:20:07.676553Z  120      2.561   
20             23          1  2024-04-06T07:20:22.597454Z  121      1.798   
21             24          1  2024-04-06T07:20:30.653422Z  122      3.157   
22             25          1  2024-04-06T07:20:38.584335Z  123      2.064   
23             26          1  2024-04-06T07:20:47.373324Z  124      2.279   
24             27          1  2024-04-06T07:20:56.013882Z  125      1.986   
25             28          1  2024-04-06T07:21:05.041334Z  126      2.779   
26             29          1  2024-04-06T07:21:13.326542Z  127      2.285   
27             30          1  2024-04-06T07:21:26.585499Z  128      2.161   
28             31          1  2024-04-06T07:21:36.448675Z  129      2.257   
29             32          1  2024-04-06T07:21:51.609330Z  130      1.950   
30             33          1  2024-04-06T07:22:00.814257Z  131      2.568   
31             34          1  2024-04-06T07:22:11.532239Z  132      1.948   
32             35          1  2024-04-06T07:22:19.535400Z  133      2.492   
33             36          1  2024-04-06T07:22:31.024236Z  134      2.294   
34             37          1  2024-04-06T07:22:39.442221Z  135      1.766   
35             38          1  2024-04-06T07:22:48.232197Z  136      2.728   
36             39          1  2024-04-06T07:22:57.112181Z  137      1.589   
37             40          1  2024-04-06T07:23:04.979727Z  138      1.718   
38             41          1  2024-04-06T07:23:16.040467Z  139      1.926   
39             42          1  2024-04-06T07:23:30.391606Z  140      3.616   
40             43          1  2024-04-06T07:23:38.880155Z  141      1.786   
41             44          1  2024-04-06T07:23:47.129119Z  142      2.086   
42             45          1  2024-04-06T07:23:56.125169Z  143      2.132   
43             46          1  2024-04-06T07:24:05.242099Z  144      1.909   
44             47          1  2024-04-06T07:24:13.550094Z  145      1.884   
45             48          1  2024-04-06T07:24:25.129225Z  146      2.438   
46             49          1  2024-04-06T07:24:33.349949Z  147      2.986   
47             50          1  2024-04-06T07:24:40.312265Z  148      1.919   
48             51          1  2024-04-06T07:24:49.948274Z  149      1.627   
49             52          1  2024-04-06T07:24:56.893132Z  150      1.829   

                                               review rule_based_sentiment  

Задача 4.     
Объедините данные, размеченные вручную, с данными, размеченными на основе правил. Объедините два подмножества размеченных данных в один набор данных, сохранив при этом соответствующую структуру и целостность.

In [ ]:
manual_labeling_data = pd.read_csv("manual_data.csv")
auto_manual_data = pd.read_csv("auto_manual.csv")

# Объединим данные
merged_data = pd.concat([manual_labeling_data, auto_manual_data])

# Сохраняем объединенные данные в файл CSV
merged_data.to_csv('subset_manual_check.csv', index=False)

In [ ]:
# Загрузим данные из файла CSV
merged_data = pd.read_csv('subset_manual_check.csv')

# Удалим ненужные столбцы
merged_data = merged_data.drop(['sentiment', 'annotation_id', 'annotator', 'created_at', 'id', 'lead_time', 'updated_at'], axis=1)

# Сохраним обновленные данные в файл CSV
merged_data.to_csv('subset_manual_check_cleaned.csv', index=False)

In [ ]:
# Загрузим очищенные данные из файла CSV
cleaned_data = pd.read_csv('subset_manual_check_cleaned.csv')

# Выведем первые несколько строк для просмотра
cleaned_data.head()

review rule_based_sentiment
0  I really liked this Summerslam due to the look...             negative
1  Not many television shows appeal to quite as m...              neutral
2  The film quickly gets to a major chase scene w...             negative
3  Jane Austen would definitely approve of this o...             positive
4  Expectations were somewhat high for me when I ...             positive

Задача 5.     
Обучите модель машинного обучения, используя объединенный набор размеченных данных. Разделите датасет на обучающий и тестовый наборы и используйте обучающий набор для обучения модели.



In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score

# Загрузим данные из файла CSV
data = pd.read_csv('subset_manual_check_cleaned.csv')

# Разделим данные на признаки (X) и целевую переменную (y)
X = data['review']
y = data['rule_based_sentiment']

# Разделим данные на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Преобразуем текстовые данные в матрицу признаков с помощью мешка слов
vectorizer = CountVectorizer()
X_train_vectors = vectorizer.fit_transform(X_train)
X_test_vectors = vectorizer.transform(X_test)

# Инициализируем и обучим модель на обучающих данных
model = MultinomialNB()
model.fit(X_train_vectors, y_train)

# Предскажем метки для тестового набора
y_pred = model.predict(X_test_vectors)




Задача 6.   
Оценить эффективность обученной модели на тестовом датасете. Используйте подходящие метрики оценки. Интерпретируйте результаты и проанализируйте эффективность модели в решении задачи разметки.

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Оценим точность модели
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# Оценим точность, полноту и F1-меру
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')

print("Precision:", precision)
print("Recall:", recall)
print("F1-score:", f1)


Accuracy: 0.6583
Precision: 0.7253283987322465
Recall: 0.6583
F1-score: 0.6420024231344822


1. Accuracy (Точность): Модель демонстрирует точность примерно 65.83%. Это означает, что примерно 65.83% примеров были классифицированы верно.

2. Precision (Точность): Модель показывает точность примерно 72.53%. Это означает, что примерно 72.53% объектов, классифицированных как положительные, действительно являются положительными.

3. Recall (Полнота): Модель демонстрирует полноту также примерно 65.83%. Это означает, что примерно 65.83% всех действительно положительных объектов были обнаружены моделью.

4. F1-score (F1-мера): F1-мера равна примерно 64.20%. F1-мера является средним гармоническим между точностью и полнотой. Значение F1-меры ближе к 1 указывает на более хороший баланс между точностью и полнотой.

Вывод:    
Эти результаты позволяют нам сделать вывод, что модель показывает неплохие результаты